From bf8100365ac066408f400315f643e1023f50608f Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Wed, 29 Jul 2020 22:27:31 +0200
Subject: [PATCH] Add support for querying boot switch position

The imx8mm-var-dart has a boot switch that we can use
to determine if we need to boot flasher rootfs or not.

This patch adds support for querying it from
the u-boot environment.

Upstream-status: Inappropriate [configuration]
Signed-off-by: Alexandru Costache <alexandru@balena.io
---
 .../imx8mm_var_dart/imx8mm_var_dart.c         | 33 +++++++++++++++++++
 cmd/Kconfig                                   |  5 +++
 cmd/Makefile                                  |  1 +
 cmd/bootswitch.c                              | 26 +++++++++++++++
 configs/imx8mm_var_dart_defconfig             |  1 +
 5 files changed, 66 insertions(+)
 create mode 100644 cmd/bootswitch.c

diff --git a/board/variscite/imx8mm_var_dart/imx8mm_var_dart.c b/board/variscite/imx8mm_var_dart/imx8mm_var_dart.c
index a99aace78a..fbf8c5949b 100644
--- a/board/variscite/imx8mm_var_dart/imx8mm_var_dart.c
+++ b/board/variscite/imx8mm_var_dart/imx8mm_var_dart.c
@@ -26,6 +26,17 @@ extern int var_setup_mac(struct var_eeprom *eeprom);
 #define UART_PAD_CTRL	(PAD_CTL_DSE6 | PAD_CTL_FSEL1)
 #define WDOG_PAD_CTRL	(PAD_CTL_DSE6 | PAD_CTL_ODE | PAD_CTL_PUE | PAD_CTL_PE)
 
+/* 0x1c6 in kernel dts */
+#define BOOT_PAD_CTRL    (PAD_CTL_DSE6 | PAD_CTL_HYS | PAD_CTL_PUE | PAD_CTL_PE)
+
+#define BOOT_SWITCH_EXT   IMX_GPIO_NR(4, 14)
+#define BOOT_SWITCH_INT   IMX_GPIO_NR(4, 17)
+
+static iomux_v3_cfg_t const boot_switch_pads[] = {
+       IMX8MM_PAD_SAI1_TXD2_GPIO4_IO14 | MUX_PAD_CTRL(BOOT_PAD_CTRL),
+       IMX8MM_PAD_SAI1_TXD5_GPIO4_IO17 | MUX_PAD_CTRL(BOOT_PAD_CTRL),
+};
+
 static iomux_v3_cfg_t const uart1_pads[] = {
 	IMX8MM_PAD_UART1_RXD_UART1_RX | MUX_PAD_CTRL(UART_PAD_CTRL),
 	IMX8MM_PAD_UART1_TXD_UART1_TX | MUX_PAD_CTRL(UART_PAD_CTRL),
@@ -225,6 +236,26 @@ static void setup_usb(void)
 }
 #endif
 
+int is_boot_switch(const char *position)
+{
+       if (!strncmp(position, "INT", 3) && gpio_get_value(BOOT_SWITCH_INT)) {
+               return 1;
+       } else if (!strncmp(position, "EXT", 3) && gpio_get_value(BOOT_SWITCH_EXT)) {
+               return 1;
+       }
+
+       return 0;
+}
+
+static void setup_bootswitch_gpios(void)
+{
+       imx_iomux_v3_setup_multiple_pads(boot_switch_pads, ARRAY_SIZE(boot_switch_pads));
+       gpio_request(BOOT_SWITCH_INT, "BOOT_INT");
+       gpio_direction_input(BOOT_SWITCH_INT);
+       gpio_request(BOOT_SWITCH_EXT, "BOOT_EXT");
+       gpio_direction_input(BOOT_SWITCH_EXT);
+}
+
 int board_init(void)
 {
 #ifdef CONFIG_FEC_MXC
@@ -235,6 +266,8 @@ int board_init(void)
 	setup_usb();
 #endif
 
+	setup_bootswitch_gpios();
+
 	return 0;
 }
 
diff --git a/cmd/Kconfig b/cmd/Kconfig
index 9518136575..018a36881c 100644
--- a/cmd/Kconfig
+++ b/cmd/Kconfig
@@ -1347,6 +1347,11 @@ endif
 
 menu "Misc commands"
 
+config CMD_BOOTSWITCH_POS
+        bool "Enable 'bootswitchpos' command"
+        help
+          This provides a way to determine boot switch position on the variscite imx8mm-var-dart
+
 config CMD_BMP
 	bool "Enable 'bmp' command"
 	depends on LCD || DM_VIDEO || VIDEO
diff --git a/cmd/Makefile b/cmd/Makefile
index 9646585052..37bcb8ff17 100644
--- a/cmd/Makefile
+++ b/cmd/Makefile
@@ -162,6 +162,7 @@ obj-$(CONFIG_CMD_ZFS) += zfs.o
 obj-$(CONFIG_CMD_DFU) += dfu.o
 obj-$(CONFIG_CMD_GPT) += gpt.o
 obj-$(CONFIG_CMD_ETHSW) += ethsw.o
+obj-$(CONFIG_CMD_BOOTSWITCH_POS) += bootswitch.o
 obj-$(CONFIG_CMD_AXI) += axi.o
 
 # Power
diff --git a/cmd/bootswitch.c b/cmd/bootswitch.c
new file mode 100644
index 0000000000..8e0c66809f
--- /dev/null
+++ b/cmd/bootswitch.c
@@ -0,0 +1,26 @@
+#include <common.h>
+#include <config.h>
+#include <command.h>
+#include <vsprintf.h>
+
+#ifdef CONFIG_IMX8MM
+#ifdef CONFIG_CMD_BOOTSWITCH_POS
+
+#define USAGE "bootswitchpos <INT/EXT>\n"
+
+int is_boot_switch(const char*);
+
+static int do_bootswitchpos(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
+{
+       if (argc < 2) {
+               printf(USAGE);
+               return 1;
+       }
+
+       return !is_boot_switch(argv[1]);
+}
+
+U_BOOT_CMD(bootswitchpos, 2, 1, do_bootswitchpos, "", "")
+
+#endif
+#endif
diff --git a/configs/imx8mm_var_dart_defconfig b/configs/imx8mm_var_dart_defconfig
index 4542b2635a..c6cdd2d439 100644
--- a/configs/imx8mm_var_dart_defconfig
+++ b/configs/imx8mm_var_dart_defconfig
@@ -68,3 +68,4 @@ CONFIG_CMD_EXT2=y
 CONFIG_CMD_EXT4=y
 CONFIG_CMD_FAT=y
 CONFIG_CMD_EXPORTENV=y
+CONFIG_CMD_BOOTSWITCH_POS=y
-- 
2.17.1

